Barline issues

MarcSabatella opened the pull request:
So far I have fixes for:

https://musescore.org/en/node/286017
https://musescore.org/en/node/280912
https://musescore.org/en/node/102821

Currently, we don't link barlines between score & parts.  If we decided we wanted to, then this would change the approach used here, but I didn't want t change anything so fundamental, so I link barlines by their staves or scores.  The commit to fix the second & third issues listed here actually fixes quite a few related problems (eg, deleting a repeat on an mmrest does nothing, also issues with barlines getting marked "non-generated" and then not responding to changes in span for the staff).  So it's actually a very significant fix.

The code as I have it is good, I think, but I'd still like to look into https://musescore.org/en/node/282696 and https://musescore.org/en/node/283366.  I don't think this will require visiting existing code, but it does mean re-thinking some things about "generated" status, so who knows.

MarcSabatella commented:
As of now, this PR incorporates fixes for all the high priority regressions (and a couple of bugs that were in 2.3.2 also) I know about and consider must-fixes for 3.1.  I didn't yet deal with the request to reinstate dragging the top barline up; it doens't seem any major loss of functionality.  Shouldn't be hard - the old code seems to still be there but commented out, and updating it to work with the new representation should be simple enough - but actually, given how things work now with everything controlled by a simple "span to next staff" flag, it seems actually a little odd to do this.

AntonioBL commented:
What about the case in which a non-generated barline (e.g. a double barline) exists in the score before creating the parts?
For example: new score; add a double barline somewhere; generate the parts (e.g. file->parts->New all->ok)
I think that currently in this case the barlines in the masterscore and in the part get linked together.

MarcSabatella commented:
Good point, they *are* linked in that case.  But that doesn't cause any trouble for my code that I can see, I basically ignore links on the barlines and do everything by following the links on the staves or the scores themselves and then locating the correct barline according to tick.  I tried and couldn't reproduce any problem.

In theory we could certainly try an approach where we link all barlines, but it could be considerably more complicated, because maintaining links on generated elements requires additional upkeep.  I suspect that is why your initial attempt crashed on undo, we would need to manage this linking in more places, be consistent about which changes get pushed to the undo stack, etc.

AntonioBL commented:
So maybe it is better to avoid linking the barlines also in that case, so that there is not an underlying difference between non-generated barlines made before and after creating the parts.

MarcSabatella commented:
Maybe, indeed.  I'm not crazy about changing something like that without good reason, but I don't like inconsistency, either.  So, I've pushed another commit that accomplishes this - skipping the link for non-generated barlines when creating the excerpt.  I expect it will break a bunch of tests, as the link info will be missing from the score file when saving a score with parts.  Not the end of the world, but I don't have an easy way to update tests myself right now, so if it's more than a few, I may back that change out and leave it for someone else to worry about later.

MarcSabatella commented:
The change broke a bunch of guitarpro tests, but not just in the linking of barlines - somehow it affected chords as well and changed the link tags for them too.  So, I'd rather not deal with that.  I have thus pushed another commit reverting the previous one.  If someone else wants to grab the previous commit https://github.com/musescore/MuseScore/pull/4843/commits/8ad480b7999d118dc71079db215ca0323fb20f96 and work on investigating why it breaks guitarpro tests, be my guest, but I do think my PR stands on its own.

I will leave my commit history for now for further review purposes, but can squash if/when the time comes. if nothing else to remove the last two commits (which cancel each other out).

MarcSabatella commented:
As far as I am concerned, this is mergeable as it, but I am now considering adding another commit or two to fix other issues.  I don't want to use a different branch because likely my changes would need to piggy-back on what I've already done here.  Just a heads up.  As far as I know, it should work equally well to merge this while I'm working on the other issues, or to wait.

MarcSabatella commented:
The last commit (adding end-start repeat back to the palette) borrows from #4812 but takes advantage of my code here for linking to parts, also other miscellaneous improvements.

AntonioBL commented:
I think the failure of GuitarPro tests when disabling linking could be due to the additional tab staff which, for these files, is created in the parts but it is not present in the main score.
From what I see for this additional staff the function Excerpt::cloneStaff is called instead of Excerpt::cloneStaves.

MarcSabatella commented:
Could be.  I'm thinking of adding the linking as you suggested originally though, mostly to facilitate adding symbols to the barlines.  I will likely do it just for non-generated barlines though.

MarcSabatella commented:
Added code to link non-generated barlines in both mmrests and linked staves/parts, whether or not the barlines existed before creating the part.  I also unlink them when reverting to generated status (eg, changing a double bar back to normal).  Everything works on undo/redo in my own testing, but we'll see what the mtests say.  Very possibly I'll need to update a few reference files.

One I have a clean Travis build on this, I believe I am done mucking with infrastructure.

MarcSabatella commented:
I recognize this is a pretty big PR.  It fixes a lot of interrelated issues and some of the changes depend on others in ways that made it easiest to implement and test with them combined.

There is one commit that implements a proposed change to how linking works, and then the next commit reverts that.  Before merging, I would be by happy to do a rebase to squash those two into the previous commit.  Other than that, though, I think the commits should remain independent, as they are actually pretty cleanly separated.

MarcSabatella commented:
I went ahead and rebased against current master and squashed out those two commits, no other changes.

In case it isn't clear, I personally consider these issues top priority for 3.1.  No single one is necessarily a show-stopper in itself, but collectively they result in the non-unfounded perception that barlines are just completely broken in MuseScore 3 right now.  For example, see https://musescore.org/en/node/286714
